### PR TITLE
[JENKINS-55048] Add support for plugin Java requirement metadata

### PR DESCRIPTION
**Pull request by daniel-beck**
See [JENKINS-55048](https://issues.jenkins-ci.org/browse/JENKINS-55048).

Downstream of https://github.com/jenkinsci/maven-hpi-plugin/pull/75 and https://github.com/jenkins-infra/backend-update-center2/pull/164.

Future enhancement might be to send the Java version with the request (or consider it already sent by the user agent header) and handle that on the server side, similar to the tiered update sites. But for now, a big red (newly bold) warning should at least get us started.

Here is how it renders:
image

(I used http://updates.jenkins.io/temporary-experimental-java11/ for testing.
`Pipeline: Supporting APIs`, aka `workflow-support` is the only one requiring Java 11+.
The others are red because they depend on it directly or transitively.
I used the following system groovy quick&dirty script to check these things:

```groovy
def showDeps(plugin, level) {
  println ("  ".multiply(level)+"${plugin.name}:${plugin.version}")
  
  plugin.getNeededDependencies().each { dep ->
    def dep1 = dep
    showDeps(dep1, level + 1);
  }
}
showDeps(Jenkins.instance.updateCenter.getPlugin('throttle-concurrents'), 0)
```
).

### Proposed changelog entries

* Add support for plugins declaring a minimum Java version in manifest and update center, showing warnings and refusing to load plugins with unsatisfied dependencies.

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests

### Desired reviewers

@jenkinsci/code-reviewers

### CODE REVIEW

File: core/src/main/java/hudson/PluginWrapper.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/PluginWrapper.java]: Document what `null` means here? Probably it also makes sense to extrapolate the core dependency to a Java version if undefined (or Just return core Java version once/if #2949 is integrated)

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/UpdateSite.java]: Needs a comment like above. I would probably move it to a static method in order to integrate with #2949

File: core/src/main/resources/hudson/PluginManager/_table.css
Comment (oleg-nenashev) [inline core/src/main/resources/hudson/PluginManager/_table.css]: line break

**Review by oleg-nenashev — COMMENTED** [review]
Generally the PR looks good, added few comments.

I would also consider adding `Maximum-Java-Version` as a part of this story while you are around. E.g. I barely believe that Maven Integration Plugin can be made Java9-compatible easily (CC @aheritier)

File: core/src/main/java/hudson/PluginWrapper.java
Comment (jglick) [reply core/src/main/java/hudson/PluginWrapper.java]: I think null is fine here: means the plugin does not specify a version.

File: core/src/main/java/hudson/PluginWrapper.java
Comment (jglick) [inline core/src/main/java/hudson/PluginWrapper.java]: More simply,

```java
return manifest.getMainAttributes().getValue("Minimum-Java-Version");
```

File: core/src/main/java/hudson/PluginWrapper.java
Comment (jglick) [inline core/src/main/java/hudson/PluginWrapper.java]: You want `java.specification.version`. The format of `java.version` is undefined.

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (jglick) [reply core/src/main/java/hudson/model/UpdateSite.java]: as above, this is the wrong property

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (jglick) [inline core/src/main/java/hudson/model/UpdateSite.java]: Huh? This should be logged as a `WARNING`.

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (jglick) [inline core/src/main/java/hudson/model/UpdateSite.java]: log

**Review by jglick — DISMISSED** [review]
Misuse of `java.version`, otherwise looks good.

**Comment by daniel-beck** [comment]
Rebased to current master with no impl changes so far.

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (daniel-beck) [reply core/src/main/java/hudson/model/UpdateSite.java]: I really don't see how this could happen, but basically copied it from the existing code, which behaves identically.

File: core/src/main/java/hudson/PluginWrapper.java
Comment (batmat) [reply core/src/main/java/hudson/PluginWrapper.java]: I think so too. And it's consistent with other similar methods here.

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (batmat) [reply core/src/main/java/hudson/model/UpdateSite.java]: I agree with @jglick this looks wrong. The comment is misleading: this will enter the catch block iff `minimumJavaVersion` is not null, and hence specified.

Then would enter the block if `new VersionNumber(minimumJavaVersion)` throws a `NumberFormatException`. So if a minimumJavaVersion **is** set AND cannot be parsed: this should very likely be reported or I'm missing something.

File: test/src/test/resources/plugins/minJavaVersion-update-center.json
Comment (batmat) [inline test/src/test/resources/plugins/minJavaVersion-update-center.json]: Hopefully, we shouldn't get too quickly to this as a baseline (?). Or maybe I should even have gone bigger.

**Comment by batmat** [comment]
@jenkinsci/java11-support please review this pull-request.

**Comment by batmat** [comment]
To test this, a simple way is the following:

```shell
# Build and run a Jenkins with this patch
docker run -ti -e ID=3016 -p 8080:8080 batmat/jenkins-pr-tester
```

Then open a browser on 8080, and change the update center URL to http://updates.jenkins.io/temporary-experimental-java11/update-center.json to check how it behaves. 
In this UC, the `workflow-support` plugin in 3.0-java11-alpha-11` requires Java 11.

**Comment by batmat** [comment]
Tested also using the `jenkins/jenkins:jdk11` Docker image, the warnings are not shown, as expected.

`docker run -ti -p 8080:8080 -v ~/path/to/repo/jenkinsci/jenkins/war/target/jenkins.war:/usr/share/jenkins/jenkins.war:ro jenkins/jenkins:jdk11`

File: test/src/test/resources/plugins/minJavaVersion-update-center.json
Comment (MarkEWaite) [inline test/src/test/resources/plugins/minJavaVersion-update-center.json]: JSON doesn't allow comments.  Does this work?

If so, why?

**Comment by daspilker** [comment]
Pinging @sghill as the new maintainer of the Gradle JPI plugin.

File: test/src/test/resources/plugins/minJavaVersion-update-center.json
Comment (batmat) [reply test/src/test/resources/plugins/minJavaVersion-update-center.json]: > Does this work?

Yes. It works, you can see the test does use it.

> If so, why?

It's history I suppose :). This is the format used on all Jenkins update center files. 
You can check for yourself in `curl -L -r 99 https://updates.jenkins.io/update-center.json`

(BTW, a few years ago were added the actual json exports, add `.actual` in between, like https://updates.jenkins.io/update-center.actual.json).

Comment (alecharp) [reply test/src/test/resources/plugins/minJavaVersion-update-center.json]: because it's not  JSON but JSONP, maybe comments are authorized in this format.

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (MRamonLeon) [inline core/src/main/java/hudson/model/UpdateSite.java]: typo: iff -> if

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (MRamonLeon) [inline core/src/main/java/hudson/model/UpdateSite.java]: same typo

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (MRamonLeon) [inline core/src/main/java/hudson/model/UpdateSite.java]: Can you set the message inside the method avoiding passing it as an argument?. It's always the same

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (MRamonLeon) [inline core/src/main/java/hudson/model/UpdateSite.java]: :laughing:  look for *iff* in code

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (MRamonLeon) [inline core/src/main/java/hudson/model/UpdateSite.java]: The comment doesn't fit the code, I think. The code is not taking into account if this plugin *isForNewerJava*, it only checks its dependencies. Or I'm missing something. I don't know whether the code is wrong, or just the comment

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (MRamonLeon) [inline core/src/main/java/hudson/model/UpdateSite.java]: The name of the method is tricky for me. Do you like instead, something like **getMinimumJavaVersionConsideringDependencies** or similar? :pray:

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (MRamonLeon) [reply core/src/main/java/hudson/model/UpdateSite.java]: gotcha

File: test/src/test/resources/plugins/minJavaVersion-update-center.json
Comment (fcojfernandez) [reply test/src/test/resources/plugins/minJavaVersion-update-center.json]: AFAIK, neither JSON nor JSONP allow comment, but depending on the library you use for parsing the data, the comments are ignored.

Comment (MarkEWaite) [reply test/src/test/resources/plugins/minJavaVersion-update-center.json]: Would you be willing to remove the comment, or place it in an unused data element?  

I think GitHub is right to flag it with red in the diffs, since the json file format does not allow comments.

If not, I'm ok with that as well.

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (batmat) [reply core/src/main/java/hudson/model/UpdateSite.java]: I tend to agree, but I used the existing naming pattern. So I think either we'd fix everything, or nothing.

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (batmat) [reply core/src/main/java/hudson/model/UpdateSite.java]: fixed, nice catch.

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (batmat) [reply core/src/main/java/hudson/model/UpdateSite.java]: Yeah, I considered it, but didn't do it because this seemed not like a big deal. Looking.

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (alecharp) [inline core/src/main/java/hudson/model/UpdateSite.java]: between this and [#763](https://github.com/jenkinsci/jenkins/pull/3016/files#diff-dd00e0527600a2b32c10fe2d5af188adR763), could we externalize (in a method) how we get the Java version? This way, we make sure that any part of the code is using the same system and when we will be able to change it by `Runtime.version()` we will have to do it in one place?

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (MRamonLeon) [reply core/src/main/java/hudson/model/UpdateSite.java]: Nothing to be consistent then.

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (MRamonLeon) [inline core/src/main/java/hudson/model/UpdateSite.java]: I think you can omit the word **it** here and in the @return section

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (MRamonLeon) [inline core/src/main/java/hudson/model/UpdateSite.java]: As the calls to this method is always the same you can even avoid passing the **minimumJavaVersion** argument.

**Review by MRamonLeon — APPROVED** [review]
Minor comments, but anyway it LGTM

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (batmat) [reply core/src/main/java/hudson/model/UpdateSite.java]: Yes, absolutely. I almost did it, but didn't want to pull myself into an out of scope change in the same PR. (_looks_ tiny, granted, but devil's in the details...). I had a look, but given Runtime.version() returns a new `Runtime.Version` class, it would not be _such_ a straightforward and forward compatible function.

I would propose to file a separate PR to group the retrievals to java.specification.version (or wrong java.version) as there are many in the core.
Likely, this could even be centralized in the versionnumber-lib.

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (batmat) [reply core/src/main/java/hudson/model/UpdateSite.java]: In *this* PR, I agree adjusting everything would be out of scope. But fixing it in a dedicated PR could make sense.

File: test/src/test/resources/plugins/minJavaVersion-update-center.json
Comment (batmat) [reply test/src/test/resources/plugins/minJavaVersion-update-center.json]: I'm going to remove it. I think it works because the parser is permissive.

For the record, I didn't write this myself. There are already many similar `.json` files in that resource directory which I copied and modified for my test case.

Comment (batmat) [reply test/src/test/resources/plugins/minJavaVersion-update-center.json]: Done @MarkEWaite

**Review by MarkEWaite — APPROVED** [review]

File: core/src/main/java/hudson/PluginWrapper.java
Comment (jglick) [inline core/src/main/java/hudson/PluginWrapper.java]: missing space

File: core/src/main/java/hudson/PluginWrapper.java
Comment (jglick) [inline core/src/main/java/hudson/PluginWrapper.java]: Please specify the format here. `1.8`? `8`? Something else?

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (jglick) [reply core/src/main/java/hudson/model/UpdateSite.java]: At least a method in `Util` for now.

File: core/src/main/java/hudson/PluginWrapper.java
Comment (jglick) [inline core/src/main/java/hudson/PluginWrapper.java]: BTW the practice has been to use `FIXME` in these comments intended to be cleaned up immediately after the merge if not in the merge commit itself, as opposed to “rainy-day” comments that could sit there for years. I guess to make it easier to grep for them.

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (jglick) [inline core/src/main/java/hudson/model/UpdateSite.java]: Avoid duplicating text between summary and `@return` tag. You can omit one or the other, or use the `@return` tag to delve into details of format, while the summary gives the overall purpose.

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (jglick) [inline core/src/main/java/hudson/model/UpdateSite.java]: Meaning of null should be documented in `@return`.

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (jglick) [inline core/src/main/java/hudson/model/UpdateSite.java]: Might have missed a doubling of `'` here. I recommend just avoiding ASCII apostrophes in message formats. Use `‘…’` or nothing.

File: core/src/main/resources/hudson/PluginManager/table.properties
Comment (jglick) [inline core/src/main/resources/hudson/PluginManager/table.properties]: _requires dependent plugins_ reads a little awkwardly—it is redundant. I might suggest something like

> Warning: this plugin requires other plugins which in turn require Java…

File: test/src/test/java/hudson/model/UpdateSiteTest.java
Comment (jglick) [inline test/src/test/java/hudson/model/UpdateSiteTest.java]: `@Issue`

File: test/src/test/resources/plugins/minJavaVersion-update-center.json
Comment (jglick) [inline test/src/test/resources/plugins/minJavaVersion-update-center.json]: This is JSONP. Either put it in a file with a `.js` extension, or actually use JSON, which IIRC has been supported for a few years now.

**Review by jglick — APPROVED** [review]

**Review by fcojfernandez — APPROVED** [review]

File: core/src/main/java/hudson/PluginWrapper.java
Comment (batmat) [reply core/src/main/java/hudson/PluginWrapper.java]: It _seems_ like there are actually `@since TODO` only currently in the codebase.

Anyway, I filed #3837 to address these more easily in the future.

File: test/src/test/resources/plugins/minJavaVersion-update-center.json
Comment (batmat) [reply test/src/test/resources/plugins/minJavaVersion-update-center.json]: I don't disagree, it just feels out of scope given there're many files like these already in the folder. But can fix in a followup if you feel any strongly.

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (batmat) [reply core/src/main/java/hudson/model/UpdateSite.java]: done

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (batmat) [reply core/src/main/java/hudson/model/UpdateSite.java]: done

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (batmat) [reply core/src/main/java/hudson/model/UpdateSite.java]: :+1: done.

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (batmat) [reply core/src/main/java/hudson/model/UpdateSite.java]: done

**Comment by batmat** [comment]
@jglick I believe I addressed your feedback. Can you please take another quick look?

Thanks!

**Comment by batmat** [comment]
BTW, anyone willing to have another look is welcome.
The additions I made since approvals are very trivial, often in javadoc only but a few confirmed approvals would be better I suppose.

Thanks!

File: core/src/main/java/jenkins/util/java/JavaUtils.java
Comment (jglick) [inline core/src/main/java/jenkins/util/java/JavaUtils.java]: Best to document this with examples. IIUC these would be

* `1.8`
* `9`
* `10`
* `11`

?

**Review by jglick — APPROVED** [review]

File: core/src/main/java/jenkins/util/java/JavaUtils.java
Comment (batmat) [reply core/src/main/java/jenkins/util/java/JavaUtils.java]: Yes. Basically [JEP-223](https://openjdk.java.net/jeps/223)

Comment (batmat) [reply core/src/main/java/jenkins/util/java/JavaUtils.java]: @jglick fix proposed in https://github.com/jenkinsci/jenkins/pull/3842

**Comment by oleg-nenashev** [comment]
It causes https://issues.jenkins-ci.org/browse/JENKINS-55562 , because there is no version normalization code in this pull request. Both `8` and `1.8` are valid version numbers, because pre-JEP-223 format stays valid for older versions

**Pull request merged**
